src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/OauthCrumbExclusion.java

```diff
@@ -0,0 +1,31 @@
+package com.atlassian.bitbucket.jenkins.internal.applink.oauth.serviceprovider.rest;
+
+import com.atlassian.bitbucket.jenkins.internal.applink.oauth.serviceprovider.OAuthRequestUtils;
+import hudson.Extension;
+import hudson.security.csrf.CrumbExclusion;
+
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+
+/**
+ * Adds exception to the CSRF protection filter for OAuth (applink) requests.
+ */
+@Extension
+public class OauthCrumbExclusion extends CrumbExclusion {
+
+    @Override
+    public boolean process(HttpServletRequest req, HttpServletResponse resp, FilterChain chain)
+            throws IOException, ServletException {
+        String pathInfo = req.getPathInfo();
+        if (isEmpty(pathInfo) || !OAuthRequestUtils.isOAuthAccessAttempt(req)) {
```

Comment:
Is the `isEmpty` required? `isOAuthAccessAttempt` is pulling out the path info, so I think the call is redundant.

Reply:
Request tokens aren't part of this solution- do we provide credentials when hitting the request token endpoint?

Reply:
I shamelessly stole the webhook one, and you're right, the `isEmpty` and the whole `pathInfo` is not needed, I'll update

Reply:
As for request tokens, it is my understanding that they do not require permissions. @alibk0rd or @goravjoshi can confirm if my understanding is corerct

Reply:
Also, I should point out, this is not about _credentials_ it is about disabling the XSRF protection for the relevant calls. Any authentication etc still needs to be handled in the relevant resources.

Reply:
So just following up mostly for clarification (as have verified this works)- this is adding an exception for all requests that match our access token endpoint. The access token methods are served through the same action as our request tokens (`TokenEndpoint`) and I was wondering why request tokens seem to be working but access tokens require an exception? I assume the TokenEndpoint action, which is not unprotected, is where the 403 is coming from.

Reply:
Not quite, this is adding an exception for the csrf protection (which is unrelated to any form of authentication/authorisation).  to requests that are deemed to be Oauth by way of checking the `isOauthAccessAttempt` return value. If it returns 'true' (it is an oAuth access attempt) the request is allowed through even if there is no CSRF headers present. It is then up to the receiving resource to do normal authentication/authorisation checks (which may of course be delegated to Jenkins via standard means). 

All this does is to say that 'if it is an oAuthAccessAttempt do not require the CSRF headers on the request' it does not do anything else, or disable anything else.